### DISCUSSION TITLE
(fb2 css) Adjacent headings

### DISCUSSION THREAD

**Discussion by mergen3107** [koreader/koreader#8162] [Q&A]
Guys, can you please help me figure out how could I remove page-break-before for an adjacent title in a fb2 book?

Here is the two headings saying `ACT 1` and `Scene 1`:
![Screenshot_20210831-154954](https://user-images.githubusercontent.com/6279855/131566988-0184ba36-d57f-4bab-8f73-a4cd30284bf5.png)

This is how they look like in the book:
![Screenshot_20210831-154925](https://user-images.githubusercontent.com/6279855/131567039-9f282b2e-84aa-4c3d-a43b-a218015f33ac.png)

So section of `Scene 1` is adjacent to `title`.

I tried:
```
title + title {
    page-break-before: avoid !important;
}
```
also with `title + .title`, `title p + title p`, to no avail.

If I do `Avoid publisher page breaks` in the style tweaks, it removes all page-breaks. I just want to remove the page break between `ACT 1` and `Scene 1`.

Thank you :)

**Comment by mergen3107** [comment]
I found the solution myself:
```
title + section title {
    page-break-before: avoid !important;
}
```

This `section title` acts as a next-level heading (à la `h2` I guess). I'll see if it breaks anything down the book.

**Reply by poire-z** [reply]
May be `section > title + section > title` if you want to avoid it to be greedy (`>` means direct descendant, while a blank/space means any descendant deep down - or ancestor deep up) - unless you want any other `title` down in your initial title+section to avoid that).
Might be candidate for inclusion in our fb2.css.

Or may be better, just avoid a page break after any section > title, whether it's text or another section > title:

`section + title { page-break-after: avoid; }`
It should have a [higher specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity) than the one that cause your break in our fb2.css::
```css
title, h1, h2 {
        page-break-before: always;    /* it should override this one */
        page-break-inside: avoid;
        page-break-after: avoid;
}
```

**Reply by Frenzie** [reply]
Or even a step up from that in specificity with something like `section > title + section:first-child > title:first-child`.

**Reply by mergen3107** [reply]
> `section > title + section > title`

out of these suggestions, only this one works the same as `title + section title`. So I'll keep it for now.

P.S. It is so handy to be able to edit Book specific tweaks right in KOReader. Is it possible to add this feature for User Style Tweak as well? Like, create new, delete, edit?

**Reply by poire-z** [reply]
Just use the Text Editor plugin for that.
We could possibly have long-press on the user style tweak CSS snippet area bring the TextEditor (which would allow edit/delete), but for "create", you'll have to go via the TextEditor.
Also, after edit, you need to restart KOReader I think.

**Reply by mergen3107** [reply]
Ah, OK! 
For existing css, it is enough to reapply them. 
For new files, yes, restart is required to find them

**Comment by mergen3107** [comment]
Is there a way to avoid widows and orphans, but only inside stanza? The existing style tweaks for widows and orphans do not work. 
In the same book, I have something like:
```
<poem>
    <stanza>
        <v>first line</v>
        <v>...</v>
        <v>...</v>
        <v>last line</v>
    </stanza>
</poem>
```

So these first and last lines are able to become orphans and widows respectively.

**Reply by mergen3107** [reply]
OK, experimented and these seem to work:
```
stanza:first-child {
    page-break-after: avoid !important;
}

stanza:last-child {
    page-break-before: avoid !important;
}
```
first one is for orphans, second is for widows

**Reply by mergen3107** [reply]
Or not, mostly it doesn't work xD
For orphans this one is better:
```
stanza > v {
    page-break-after: avoid !important;
}
```

I'll try to make for the last one as well.

**Reply by mergen3107** [reply]
That one was also not perfect. Much better are those:
```
stanza > v:first-child {
    page-break-after: avoid !important;
}
```
and
```
stanza > v:last-child {
    page-break-before: avoid !important;
}
```

**Reply by mergen3107** [reply]
Is there a way to apply this 
```
stanza > v:first-child {
    page-break-after: avoid !important;
}
```
but only if this is not the only line? For example, if at the end of the page there are three "poems" in the row, each having exactly one line `v`, they all got pushed to the next page, leaving about half page empty.

**Reply by mergen3107** [reply]
This is what I mean:
Without the tweak
![Screenshot_20210901-203009](https://user-images.githubusercontent.com/6279855/131762831-8379c7af-7f42-4dfc-9543-fb819aec300c.png)

With the tweak
![Screenshot_20210901-203018](https://user-images.githubusercontent.com/6279855/131762848-d22b1077-5654-4377-a50d-b909ab5ade6e.png)

The names before `poem`s are `p`s, which is attached to poems with:
```
p + poem {
    page-break-before: avoid;
    margin-top: 0;
}
```

**Reply by mergen3107** [reply]
I tried this, but it seems not to work at all:
```
stanza > v:first-child:not(:only-child) {
    page-break-after: avoid !important;
}
```

**Reply by poire-z** [reply]
crengine doesn't support `:not()` - but you can override/reset it but declaring :only-child after :first-child:
```css
stanza > v:first-child {
    page-break-after: avoid !important;
}
stanza > v:only-child {
    page-break-after: auto !important;
}
```
Btw, in FB2, I think you don't need any `!important` at all: these tweaks are added after fb2.css, and after them, there is no embedded/publisher CSS in FB2 (I think), so nothing that could eventually override these and that `!important` would prevent.

For reference, the CSS bits are the strings in blue in https://github.com/koreader/crengine/blob/master/crengine/src/lvstsheet.cpp

**Reply by mergen3107** [reply]
> crengine doesn't support `:not()` - but you can override/reset it but declaring :only-child after :first-child:
> 
> ```css
> stanza > v:first-child {
>     page-break-after: avoid !important;
> }
> stanza > v:only-child {
>     page-break-after: auto !important;
> }
> ```
> 
> Btw, in FB2, I think you don't need any `!important` at all: these tweaks are added after fb2.css, and after them, there is no embedded/publisher CSS in FB2 (I think), so nothing that could eventually override these and that `!important` would prevent.
> 
> For reference, the CSS bits are the strings in blue in https://github.com/koreader/crengine/blob/master/crengine/src/lvstsheet.cpp

Rarely there are style tweaks, but it is so random that I don’t want to test it :D 

Thanks for suggestions!
